utils/fixed_decimal/src/decimal.rs

```diff
@@ -1946,51 +1932,16 @@ impl FixedDecimal {
                     return Err(Error::Limit);
                 }
             }
-            DoublePrecision::Magnitude(mag, mode) => {
-                if mag > lowest_magnitude {
-                    let round_by = (mag - lowest_magnitude) as u16;
-
-                    if round_by as usize <= n_digits {
-                        decimal.round_trailing_digits(round_by, mode)?;
-                    } else {
-                        // If we need to round by more digits than rounding can ever produce
-                        // the number is zero
-                        decimal = Default::default();
-                    }
-                }
-                if mag < 0 {
-                    // If the target magnitude was negative, make
-                    // sure we update the lower magnitude to match it
-                    decimal.lower_magnitude = mag;
-                } else if decimal.lower_magnitude < 0 {
-                    // If the target magnitude was positive,
-                    // we may have truncated digits off the right
-                    // side, reset the lower magnitude
-                    decimal.lower_magnitude = 0;
-                }
+            DoublePrecision::Magnitude(mag) => {
+                decimal.half_even(mag);
             }
-            DoublePrecision::SignificantDigits(sig, mode) => {
-                let sig = sig as usize;
+            DoublePrecision::SignificantDigits(sig) => {
                 if sig == 0 {
                     return Err(Error::Limit);
                 }
-                if sig < n_digits {
-                    let round_by = (n_digits - sig) as u16;
-                    decimal.round_trailing_digits(round_by, mode)?;
-                    // It may have rounded up by one
-                    debug_assert!(decimal.digits.len() <= sig as usize);
-                }
-                let target_lowest_magnitude = decimal.magnitude - sig as i16 + 1;
-                if target_lowest_magnitude <= 0 {
-                    // If the target magnitude was negative, make sure we update
-                    // the lower magnitude to match it
-                    decimal.lower_magnitude = target_lowest_magnitude;
-                } else if decimal.lower_magnitude < 0 {
-                    // If the target magnitude was positive,
-                    // we may have truncated digits off the right side
-                    // of the decimal point, reset the lower magnitude
-                    decimal.lower_magnitude = 0;
-                }
+
+                let position = decimal.magnitude.wrapping_sub((sig - 1) as i16);
```

Comment:
Suggestion: everything should be in bounds, so you can just write

```suggestion
                let position = decimal.magnitude - (sig as i16) + 1;
```

Reply:
what will happen if the `sig == 100` and `decimal.magnitude == i16::MIN`. I prefer to use `wrapping_sub` instead of having unknown behaviour.

Reply:
In this narrow situation, `decimal.magnitude` can't exceed the limit of a double, which is 308 to -324.

https://en.wikipedia.org/wiki/Double-precision_floating-point_format